src/Util/HTTPSignature.php

```diff
@@ -405,4 +264,178 @@ private static function decryptSigheader($header, $prvkey = null)
+
+		$signature = base64_encode(Crypto::rsaSign($signed_data, $owner['uprvkey'], 'sha256'));
+
+		$headers[] = 'Signature: keyId="' . $owner['url'] . '#main-key' . '",algorithm="rsa-sha256",headers="(request-target) content-length digest host",signature="' . $signature . '"';
+
+		$headers[] = 'Content-Type: application/activity+json';
+
+		Network::post($target, $content, $headers);
+		$return_code = BaseObject::getApp()->get_curl_code();
+
+		logger('Transmit to ' . $target . ' returned ' . $return_code);
+	}
+
+	/**
+	 * @brief Gets a signer from a given HTTP request
+	 *
+	 * @param $content
+	 * @param $http_headers
+	 *
+	 * @return signer string
+	 */
+	public static function getSigner($content, $http_headers)
+	{
+		$object = json_decode($content, true);
+
+		if (empty($object)) {
+			return false;
+		}
+
+		$actor = JsonLD::fetchElement($object, 'actor', 'id');
+
+		$headers = [];
+		$headers['(request-target)'] = strtolower($http_headers['REQUEST_METHOD']) . ' ' . $http_headers['REQUEST_URI'];
+
+		// First take every header
+		foreach ($http_headers as $k => $v) {
+			$field = str_replace('_', '-', strtolower($k));
+			$headers[$field] = $v;
+		}
+
+		// Now add every http header
+		foreach ($http_headers as $k => $v) {
+			if (strpos($k, 'HTTP_') === 0) {
+				$field = str_replace('_', '-', strtolower(substr($k, 5)));
+				$headers[$field] = $v;
+			}
+		}
+
+		$sig_block = self::parseSigHeader($http_headers['HTTP_SIGNATURE']);
+
+		if (empty($sig_block) || empty($sig_block['headers']) || empty($sig_block['keyId'])) {
+			return false;
+		}
+
+		$signed_data = '';
+		foreach ($sig_block['headers'] as $h) {
+			if (array_key_exists($h, $headers)) {
+				$signed_data .= $h . ': ' . $headers[$h] . "\n";
+			}
+		}
+		$signed_data = rtrim($signed_data, "\n");
+
+		if (empty($signed_data)) {
+			return false;
+		}
+
+		$algorithm = null;
+
+		if ($sig_block['algorithm'] === 'rsa-sha256') {
+			$algorithm = 'sha256';
+		}
+
+		if ($sig_block['algorithm'] === 'rsa-sha512') {
+			$algorithm = 'sha512';
+		}
+
+		if (empty($algorithm)) {
+			return false;
+		}
+
+		$key = self::fetchKey($sig_block['keyId'], $actor);
+
+		if (empty($key)) {
+			return false;
+		}
+
+		if (!Crypto::rsaVerify($signed_data, $sig_block['signature'], $key['pubkey'], $algorithm)) {
+			return false;
+		}
+
+		// Check the digest when it is part of the signed data
+		if (in_array('digest', $sig_block['headers'])) {
+			$digest = explode('=', $headers['digest'], 2);
+			if ($digest[0] === 'SHA-256') {
+				$hashalg = 'sha256';
+			}
+			if ($digest[0] === 'SHA-512') {
+				$hashalg = 'sha512';
+			}
+
+			/// @todo add all hashes from the rfc
+
+			if (!empty($hashalg) && base64_encode(hash($hashalg, $content, true)) != $digest[1]) {
+				return false;
+			}
+		}
+
+		// Check the content-length when it is part of the signed data
+		if (in_array('content-length', $sig_block['headers'])) {
+			if (strlen($content) != $headers['content-length']) {
```

Comment:
Are you counting bytes or characters?

Reply:
Hopefully bytes. But this - of course - depends upon the implementations on the other side. By now I had very few signature failures, so this seems to work well.

Reply:
`strlen` is correct for bytes, I was just checking for sanity.